content/blog/tidymodels-2023-q4/index.Rmd

```diff
@@ -0,0 +1,179 @@
+
+The latest release of recipes features an overhaul of the warnings and error messages to use the [cli](https://cli.r-lib.org/) package. With this, we are starting the project of providing more information signaling when things don't go well.
+
+The first type of issue we now signal for is when you try to normalize data that contains non-numeric numbers such as `NA` or `Inf`. These can sneak in for several reasons, and before this release, it happened silently. Below we are creating a recipe using the `ames` data set, and before we normalize we are taking the logarithms root of all variables that pertain to square footage. 
+
+```{r}
+rec <- recipe(Sale_Price ~ ., data = ames) |>
+  step_log(contains("SF")) |>
+  step_normalize(all_numeric_predictors()) |>
+  prep()
+```
+
+We now get a warning that something happened. Telling us that it encountered `Inf` or `-Inf`. Knowing that we can go back and investigate what went wrong. If we exclude `step_normalize()` and `bake()` the recipe, we see that a number of `-Inf` appears
+
+```{r}
+recipe(Sale_Price ~ ., data = ames) |>
+  step_log(contains("SF")) |>
+  prep() |>
+  bake(new_data = NULL, contains("SF")) |>
+  glimpse()
+```
+
+Looking at the bare data set, we notice that the `-Inf` all appear where there are `0`, which makes sense since `log(0)`.
+
+```{r}
+ames |>
+  select(contains("SF")) |>
+  glimpse()
+```
+
+Knowing that it was `0` that caused the problem, we can set an `offset` to avoid taking `log(0)`.
+
+```{r}
+rec <- recipe(Sale_Price ~ ., data = ames) |>
+  step_log(contains("SF"), offset = 0.5) |>
+  step_normalize(all_numeric_predictors()) |>
+  prep()
+```
+
+These warnings appear in `step_scale()`, `step_normalize()`, `step_center()` or `step_range()`.
+
+## Better error messages in recipes
+
+Another problem that happens a lot when using recipes, is accidentally selecting variables that have the wrong types. Previously this caused the following error:
+
+```{r}
+#| eval: false
+recipe(Sale_Price ~ ., data = ames) |>
+  step_dummy(starts_with("Lot_")) |>
+  prep()
+#> Error in `step_dummy()`:
+#> Caused by error in `prep()`:
+#> ! All columns selected for the step should be string, factor, or ordered.
+```
+
+In the newest release, it will detail the offending variables and what was wrong with them.
+
+```{r}
+#| error: true
+recipe(Sale_Price ~ ., data = ames) |>
+  step_dummy(starts_with("Lot_")) |>
+  prep() |>
+  bake()
+```
+
+## Coming Attractions
+
+In the next month or so we are planning a cascade of CRAN releases. There is a lot of new functionality coming your way, especially in the tune package. 
+
+A number of our packages will (finally) be able to cohesively fit, evaluate, tune, and predict models for event times (a.k.a., [survival analysis](https://en.wikipedia.org/wiki/Survival_analysis)). If you don't do this type of work, you might not notice the new capabilities. However, if you do, tidymodels will be able to do a lot more for you. 
```

Comment:
```suggestion
A number of our packages will (finally) be able to cohesively fit, evaluate, tune, and predict models for event times (a.k.a., [survival analysis](https://en.wikipedia.org/wiki/Survival_analysis)). If you don't do this type of work, you might not notice the new capabilities. However, if you do, tidymodels will be able to do a lot more for you. 

We've also implemented a number of features related to model fairness. These tools allow tidymodels users to identify when machine learning models behave unfairly towards certain groups of people, and will also be included in the upcoming releases of tidymodels packages in Q1.
```

Reply:
I understand that this wasn't as much of a technical feat as survival analysis, but seems worth calling out if we're referring to coming attractions. :)

Reply:
Sorry., I should have added that. I've got survival fever.